File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/TimeDiffFunction.java
package cz.fb.manaus.reactor.profit.progress.function;

import cz.fb.manaus.core.model.SettledBet;
import org.springframework.stereotype.Component;

import java.time.temporal.ChronoUnit;
import java.util.OptionalDouble;

@Component
public class TimeDiffFunction implements ProgressFunction {

    @Override
    public OptionalDouble apply(SettledBet bet) {
        var placed = bet.getPlacedOrActionDate().toInstant();
        var actionDate = bet.getBetAction().getActionDate().toInstant();
        return OptionalDouble.of(actionDate.until(placed, ChronoUnit.SECONDS));
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/TimeDiffFunction.kt
package cz.fb.manaus.reactor.profit.progress.function

import cz.fb.manaus.core.model.RealizedBet
import org.springframework.stereotype.Component
import java.time.temporal.ChronoUnit

@Component
class TimeDiffFunction : ProgressFunction {

    override fun invoke(bet: RealizedBet): Double? {
        val placed = bet.settledBet.placed
        val actionDate = bet.betAction.time
        return when {
            placed != null -> actionDate.until(placed, ChronoUnit.SECONDS).toDouble()
            else -> null
        }
    }

}
